### PR TITLE
AO-20158-Package-Install-Tests

### PR DESCRIPTION
**Pull request by ronilan**
This pull request is the **last** in a series of Pull Requests to set up GitHub Actions (for this repo and the upstream one).

### Overview
Pull Request adds a Verify (`verify.yml`) workflow that runs after the Release workflow has completed.
It provides confidence that a released package can successfully install and auto instrument apps on a wide variety of node and operatiing system combinations.

Workflow will load the [Target Group from the bindings repo](https://github.com/appoptics/appoptics-bindings-node/blob/master/.github/config/target-group.json) and for each image in group:
- Install  Appoptics agent (either the released version or the one specified manually).
- Check that installation of binary completed successfully (either from prebuilt tarball or from source).
- Start an instrumented http server to confirm auto instrumentation against staging is working.
- Run a simulation http client to generate data on staging backend.
- Start an instrumented http server to confirm auto instrumentation against production is working.

The Verify workflow can also be triggered manually on a specific version.

### Commits:

First commit (0f4bdf8) adds a simple utility client server duo to GithHub Actions. The server is a simple, Appoptics instrumented, http server. The client makes paced http requests (GET, POST, PATCH, DELETE) to the server resulting in trace data being sent to the Appoptics collector (staging). 

Second commit (33965d0) modified both client and server to accommodate various scenarios when running in GitHub Actions. Server emits error if not instrumented. Client emits error if fails to connect to server after set number of trys. Simple it is no more...

Third commit (f748536) added the Verify workflow.

### Results

The workflow hits both staging ad production backends for visual inspection

#### Staging:

Screen Shot 2021-08-26 at 10 13 49 AM

Note - clicking in shows 81 traced requests (initial GET + 20 * GET/POST/PATCH/DELETE) for all.

#### Production:

Screen Shot 2021-08-26 at 10 13 59 AM

Note - there are no traced request, only `readyToSample` check.

### Notes:
- The client server duo is a private package with own package.json. It has 2 widely used dependencies plus the appoptics agent itself that always gets version specific install. No real maintenance needed unless (far) future fails happen.
- Workflow takes about 3:30 minutes to run and 1:15 hours in commutative run time (billable).
- The workflow uses `workflow_run` trigger. It will run after Release completes successfully, but won't run if it fails.
- For simplicity, the server of the client server duo, runs on the runner rather than own container.
- When running manually, fails should be expected, depending on what stage in a release process we are at. To illustrate: if workflow is ran on `appoptics-apm@latest` right now all the node 16 non-builder images (e.g `node:16-alpine3.11`) will fail while node 16 builder images (e.g. `node:16-buster`) will pass. This is because the prerelease version of the bindings already supports node 16 with prebuilts but the latest published version of the agent uses a prior version of the bindings that does not.

### CODE REVIEW

File: .github/workflows/verify.yml [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+          image=${{ matrix.image  }}
+          pre=ghcr.io/appoptics/appoptics-bindings-node/
+          export APPOPTICS_SERVICE_KEY=${{ secrets.AO_TOKEN_STG }}:${image#"$pre"}
+          node server.js --preflight
+        working-directory: .github/utils
+
+      # results of step should be visually examined in https://my-stg.appoptics.com/
+      - name: Start Server & use Client
+        # own images name are full url. remove that to make a nice service name
+        # use client multiple times to make an easily noticeable mark on UI
```
Comment (cheempz) [inline .github/workflows/verify.yml:100]: Nit: these comments could be even more helpful on first encounter of service name prettifying, the `Preflight Staging Server` step.

**Review by cheempz — COMMENTED** [review]
💯 awesome finishing touch @ronilan!  Despite your comment that 
> Simple it is no more...

this is still a simple and clear workflow IMHO to get confidence on the release :)  I left some comments, the main one being whether we need to set the agent version in the util package.json.

Also, are you concerned that generating traces against the Production environment would cause issues?  It should be OK, so we should enable the tracing part against Production too.

File: .github/workflows/verify.yml [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+      - name: Load target group data
+        id: set-matrix
+        # json load script is does variable substitution based on repo where workflow runs
+        # this workflow runs in the agent repo thus needs to "tweak" environment for correct load.
+        run: |
+          export GITHUB_REPOSITORY=appoptics/appoptics-bindings-node
+          .github/scripts/matrix-from-json.sh .github/config/target-group.json
+
+  install-trae:
```
[the remark this answers is not in the document]
Comment (ronilan) [reply .github/workflows/verify.yml:37]: Fixed.

File: .github/utils/package.json [hunk 3]
```diff
@@ -0,0 +1,14 @@
+{
+  "name": "appoptics-apm-client-server",
+  "version": "0.0.1",
+  "private": true,
+  "description": "A simple client server duo where the server is appoptics instrumented.",
+  "main": "server.js",
+  "author": "Ron Ilan",
+  "license": "Apache-2.0",
+  "dependencies": {
+    "appoptics-apm": "^10.0.1",
```
[the remark this answers is not in the document]
Comment (ronilan) [reply .github/utils/package.json:10]: The server is dependent on `appoptics-apm` so it formally should be there, but we can make an exception. Removed.

File: .github/workflows/verify.yml (hunk 1 above)
Comment (ronilan) [reply .github/workflows/verify.yml:100]: Done.

**Comment by ronilan** [comment]
> Also, are you concerned that generating traces against the Production environment would cause issues? It should be OK, so we should enable the tracing part against Production too.

Added test against production as per offline discussion.

**Review by cheempz — APPROVED** [review]
LGTM, thanks for the revisit @ronilan!

**Pull request merged**